src/storage/src/source/postgres.rs

```diff
@@ -429,17 +430,26 @@ async fn postgres_replication_loop_inner(
                 );
             }
             Err(ReplicationError::Indefinite(e)) => {
-                // TODO: In the future we probably want to handle this more gracefully,
-                // but for now halting is the easiest way to dump the data in the pipe.
-                // The restarted storaged instance will restart the snapshot fresh, which will
-                // avoid any inconsistencies. Note that if the same lsn is chosen in the
-                // next snapshotting, the remapped timestamp chosen will be the same for
-                // both instances of storaged.
-                halt!(
+                warn!(
                     "replication snapshot for source {} failed: {}",
-                    &task_info.source_id,
-                    e
+                    &task_info.source_id, e
                 );
+                // If the channel is shutting down, so is the source.
+                let _ = task_info
+                    .sender
+                    .send(InternalMessage::Status(HealthStatusUpdate {
+                        update: HealthStatus::StalledWithError(e.to_string()),
+                        // TODO: In the future we probably want to handle this more gracefully,
+                        // but for now halting is the easiest way to dump the data in the pipe.
+                        // The restarted storaged instance will restart the snapshot fresh, which will
+                        // avoid any inconsistencies. Note that if the same lsn is chosen in the
+                        // next snapshotting, the remapped timestamp chosen will be the same for
+                        // both instances of storaged.
+                        should_halt: true,
+                    }))
+                    .await;
+
+                future::pending().await
```

Comment:
That looks strange, can't we return from the function instead?

Reply:
IDK, maybe, but I'm skeptical!

IIUC returning here will cause the downstream poller to return `Finished`, and I'm worried about that eg. advancing the output collection's upper to the empty frontier. On the other hand, just having the source be `Pending` until the halt is triggered seems much safer.